Implementing -f to flip X and Y

Guillaume-Fgt opened the pull request:
Hi Jon, first thank you to have streamed the implementation of this repo. Really interesting.

I worked on implementing the -f flip feature.

- added parsing of -f flag in `args.rs`
- in `data.rs`, I made Data responsible to provide its min-max values to Frame. For this, I created two helper fns to simplify the if-else based on flipped flag. I added the necessary if-else statements inside  fn draw_into
- in `main.rs`, I added the check for the flip option. I moved the creation of the data and the log part in its own functions (it helped me to better understand the code and the steps involved but maybe not a good idea?!).

About the cdf transform part, does it make sense to flip it or should we raise an `eyre::bail!` if --f & -cdf are present in the args parsing?

jonhoo commented:
@Guillaume-Fgt Sorry I never got back to you on this. How come you closed it? As for your question, I think it's probably fine, at least for now, to error on `-f --cdf` :+1:

Guillaume-Fgt commented:
Hi Jon, no problem, I know you are busy. I closed it because I felt I did too much refactoring and changes in the original code for a single PR. I want to make the life of maintainers as easy as possible. 

It seems ok for you, so I reopened it. But feel free to ask if you want me to make small separate PRs for each step or avoid some refactoring choices.

Guillaume-Fgt commented:
Some explanations about my commits:

* [flip parsing added](https://github.com/jonhoo/dings/pull/5/commits/72c09a9dea5922a7f263f3062ade35bdeac63bc2) ->parsing -f from command line
* [Data should be responsible for supplying its min_max vals, not Frame](https://github.com/jonhoo/dings/pull/5/commits/d7a2d7bc78b69867fa0bcc89daf26517f7d66974)->subjective choice. Tell, don't ask principle. Frame doesn't need to know detailed implementation of Data but just asks for what it needs to work.
* [created separated funcs to get min_max from vec & vec<vec>](https://github.com/jonhoo/dings/pull/5/commits/04698a51595fbd88c840956e5e02eab826d70c40)->when flipping axes, the xs and ys are flipped. In the original code, xs is Vec, ys is Vec< Vec > with hard coded min max logic. Therefore, to implement flip, I needed to put this in separate functions.
* [extracted data creation into a separate func](https://github.com/jonhoo/dings/pull/5/commits/cfb3356e6b21bcba5b866ec81b600c9cf584859b)->subjective choice of making main func smaller.
* [refactor min_max funcs for data](https://github.com/jonhoo/dings/pull/5/commits/a13df3570ee6eac7b4e5e62ca57588244942522f)->subjective choice of making min max funcs return min and max systematically instead of providing a comparator and call it two times.(I also added -f flip logic in main function at the same time, unrelated and should have been own PR).
* [moved min_y & range_y inside 'if cdf' in main.rs](https://github.com/jonhoo/dings/pull/5/commits/a8b1c54dce0bf2a7bc35f1c0908e853365baa3b9)->`min_y` & `range_y` are only used in cdf case. So I moved those variables inside `if cdf` statement.
* [moved log calcs into a separate fn in main](https://github.com/jonhoo/dings/pull/5/commits/295f0a3bb0b7edaeb121be6710309cf86f3b0380)->subjective choice of making main func smaller.
* [data get_min_max handles flip flag](https://github.com/jonhoo/dings/pull/5/commits/8f2a92c2d041c4cd5b46525d0436b66f5e6ca53e)->implementing flip logic in my min_max func.
* [first time flip is working!](https://github.com/jonhoo/dings/pull/5/commits/8d167ded5c8af8d4caab6183fea093b61d975484)->added the logic in `data.rs` to flip. This was the trickiest part and the gist of this PR with the next one.
* [data draw_into refactoring](https://github.com/jonhoo/dings/pull/5/commits/c05e559f40711d37140ea0a45becedb3e7ab77e2)->after testing several times with other flags too, I found errors and modified the logic of the `flip` flag.

Guillaume-Fgt commented:
The state of my current implementation.

Imagine you have an input like this:

> 1 2 3
> 4 5 6

Without args, you would have the points

|@|*|^|
|---|---|---|
|(0,1)|(0,2)|(0,3)|
|(1,4)|(1,5)|(1,6)|

![Untitled](https://github.com/user-attachments/assets/fa66cab4-ca0e-4a70-b037-85beefb883d3)

With the flip flag:

|@|*|^|
|---|---|---|
|(1,0)|(2,0)|(3,0)|
|(4,1)|(5,1)|(6,1)|

![Untitled](https://github.com/user-attachments/assets/9c7409d5-0d99-40cc-9c4f-7f148693d811)

With the -x tag without flip flag:

|@|*|
|---|---|
|(1,2)|(2,3)|
|(4,5)|(4,6)|

![Untitled](https://github.com/user-attachments/assets/01930ed7-7349-4476-ae9c-d2a3545aabbf)

With -x and -f flags

|@|*|
|---|---|
|(2,1)|(3,1)|
|(5,4)|(6,4)|

![Untitled](https://github.com/user-attachments/assets/a3e4b906-8a7e-4ea9-bc02-f9ccd7ed6bda)

jonhoo commented:
Reading through the diff now, I wonder if it wouldn't be easier to simply swap `x` and `y` when reading the data in, and let all the plotting logic remain the same? What do you think?
